### PR TITLE
Scalar estimators allow for the reduction over many output values (i.…

### PR DESCRIPTION
**Pull request by josephdviviano**
…e., the output of the nn.Module does not need to be a scalar, because the Estimator will apply a reduction to the final output if required).

I am not in love with the current code organization of `modules.py` -- there is some duplication, but I am thinking that a bigger refactoring effort might be en route and perhaps we should wait to optimize. Open to feedback on this!

### CODE REVIEW

File: src/gfn/modules.py [hunk 1]
```diff
@@ -306,6 +394,13 @@ def __init__(
             preprocessor=preprocessor,
             is_backward=is_backward,
         )
+        reduction_fxns = {
+            "mean": torch.mean,
+            "sum": torch.sum,
+            "prod": torch.prod,
+        }
```
Comment (younik) [inline src/gfn/modules.py:401]: here you can use the global constant, if you follow the previous comment

File: src/gfn/modules.py [hunk 2]
```diff
@@ -134,9 +134,71 @@ def to_probability_distribution(
...
+        preprocessor: Preprocessor | None = None,
+        is_backward: bool = False,
+        reduction: str = "mean",
+    ):
+        super().__init__(module, preprocessor, is_backward)
+        reduction_fxns = {
+            "mean": torch.mean,
+            "sum": torch.sum,
+            "prod": torch.prod,
+        }
```
Comment (younik) [inline src/gfn/modules.py:172]: this is constant, for convention it should go outside with upper case name

**Review by younik — APPROVED** [review]
Looks good! I added a comment for possible refactoring.

File: src/gfn/modules.py [hunk 3]
```diff
@@ -134,9 +134,71 @@ def to_probability_distribution(
...
+        preprocessor: Preprocessor object that transforms raw States objects to tensors
+            that can be used as input to the module. Optional, defaults to
+            `IdentityPreprocessor`.
+        module: The module to use. If the module is a Tabular module (from
+            `gfn.utils.modules`), then the environment preprocessor needs to be an
+            `EnumPreprocessor`.
+        preprocessor: Preprocessor from the environment.
+        _output_dim_is_checked: Flag for tracking whether the output dimenions of
+            the states (after being preprocessed and transformed by the modules) have
+            been verified.
```
Comment (hyeok9855) [inline src/gfn/modules.py:157]: This needs to be updated accordingly, e.g., add `is_backward` and `reduction` and remove `_output_dim_is_checked`.

File: src/gfn/modules.py [hunk 4]
```diff
@@ -290,13 +352,39 @@ def forward(self, states: States, conditioning: torch.tensor) -> torch.Tensor:
...
+            `IdentityPreprocessor`.
+        module: The module to use. If the module is a Tabular module (from
+            `gfn.utils.modules`), then the environment preprocessor needs to be an
+            `EnumPreprocessor`.
+        preprocessor: Preprocessor from the environment.
+        reduction_fxn: the selected torch reduction operation.
+        _output_dim_is_checked: Flag for tracking whether the output dimensions of
+            the states (after being preprocessed and transformed by the modules) have
+            been verified.
+    """
```
Comment (hyeok9855) [inline src/gfn/modules.py:378]: DITTO

File: src/gfn/modules.py [hunk 5]
```diff
@@ -134,9 +134,71 @@ def to_probability_distribution(
 
 
 class ScalarEstimator(GFNModule):
+    r"""Class for estimating scalars such as LogZ.
```
Comment (hyeok9855) [inline src/gfn/modules.py:137]: Note that `logZ` for unconditional TB is usually modeled with a single learnable parameter (`nn.Parameter`).
Should we consider modifying `ScalarEstimator` to support this kind of behavior?

File: src/gfn/samplers.py [hunk 6]
```diff
@@ -35,7 +35,11 @@ def sample_actions(
         save_estimator_outputs: bool = False,
         save_logprobs: bool = True,
         **policy_kwargs: Any,
-    ) -> Tuple[Actions, torch.Tensor | None, torch.Tensor | None,]:
+    ) -> Tuple[
+        Actions,
+        torch.Tensor | None,
+        torch.Tensor | None,
+    ]:
```
Comment (hyeok9855) [inline src/gfn/samplers.py:42]: Removing the last `,` will make this one line.

File: src/gfn/modules.py [hunk 7]
```diff
@@ -134,9 +134,71 @@ def to_probability_distribution(
...
+            "sum": torch.sum,
+            "prod": torch.prod,
+        }
+        assert reduction in reduction_fxns
+        self.reduction_fxn = reduction_fxns[reduction]
+
     def expected_output_dim(self) -> int:
         return 1
 
+    def forward(self, input: States | torch.Tensor) -> torch.Tensor:
```
Comment (hyeok9855) [inline src/gfn/modules.py:179]: In which case is the input `torch.Tensor`?

**Review by hyeok9855 — COMMENTED** [review]
Looks good! I left a few questions and comments below.

File: src/gfn/modules.py (hunk 5 above)
Comment (josephdviviano) [reply src/gfn/modules.py:137]: The GFNs themselves support this directly (you do not need to pass an estimator at all, instead you just pass a float for `Z`).

File: src/gfn/modules.py (hunk 7 above)
Comment (josephdviviano) [reply src/gfn/modules.py:179]: Yes I was looking at this and not entirely sure. It might be in the case of conditioning, where we currently don't have any sort of container, conditioning is done with a raw `Tensor`.

File: src/gfn/modules.py (hunk 1 above)
Comment (josephdviviano) [reply src/gfn/modules.py:401]: Thanks - I've done this

File: src/gfn/modules.py (hunk 5 above)
Comment (hyeok9855) [reply src/gfn/modules.py:137]: This comment is because of `such as LogZ` in the docstring!

File: src/gfn/modules.py (hunk 7 above)
Comment (hyeok9855) [reply src/gfn/modules.py:179]: Oh, it should be conditioning (e.g., conditional log Z(c)).
However, it might be a bit confusing whether to use `ConditionalScalarEstimator` or `ScalarEstimator` to model log Z(c).

Comment (josephdviviano) [reply src/gfn/modules.py:179]: Well of note, ScalarEstimators are used for more than just logZ, but in this case, I see it like this:

- LogZ can be a single parameter.
- LogZ can be estimated using a neural network - in this case, the output of the network can actually be multiple items that are averaged over.
- LogZ can be conditionally estimated using a neural network  - in this case, the output of the network can actually be multiple items that are averaged over.

From an optimization POV, sometimes having logZ only be estimated by a single parameter can cause problems (i.e., the gradients push the number around a lot), so using a neural network helps.

I agree we could make it clearer though -- I am open to suggestions.

Comment (josephdviviano) [reply src/gfn/modules.py:179]: `ConditionalScalarEstimator` is used to take in both the State and the Conditioning, i.e., it's a two-headed estimator. I think this is the normal conditioning case.

File: src/gfn/modules.py (hunk 4 above)
Comment (josephdviviano) [reply src/gfn/modules.py]: thanks :)

File: src/gfn/modules.py (hunk 5 above)
Comment (josephdviviano) [reply src/gfn/modules.py:146]: I'm not entirely sure what would be most clear here but I'm open to suggestions.

Comment (hyeok9855) [reply src/gfn/modules.py:146]: Why not just state flow functions of DB/SubTB??

**Review by hyeok9855 — APPROVED** [review]
Approved!
